### PR TITLE
[TR] PIM-4443: Fix quick export

### PR DESCRIPTION
**Pull request by willy-ahva**
| Q                  | A
| ------------------ | ---
| Migration script?  | n
| Behats             | running
| Specs & Static     | yes
| Changelog updated  | yes

### CODE REVIEW

File: spec/Pim/Bundle/TransformBundle/Normalizer/Flat/ProductNormalizerSpec.php [hunk 1]
```diff
@@ -82,6 +112,8 @@ function it_normalizes_product(
                 'family'     => 'shoes',
                 'groups'     => 'group1, group2, variant_group_1',
                 'categories' => 'nice shoes, converse',
+                012345       => "Text attribute with integer code and a zero",
```
Comment (sumbobyboys) [inline spec/Pim/Bundle/TransformBundle/Normalizer/Flat/ProductNormalizerSpec.php]: single quote please

File: spec/Pim/Bundle/TransformBundle/Normalizer/Flat/ProductNormalizerSpec.php [hunk 2]
```diff
@@ -72,9 +74,37 @@ function it_normalizes_product(
...
+                12345 => "Text attribute with integer code"
+            ]);
+        $serializer
+            ->normalize(
+                $intCodeProdValueWithZero,
+                'flat',
+                Argument::any()
+            )
+            ->willReturn([
+                012345 => "Text attribute with integer code and a zero"
```
Comment (sumbobyboys) [inline spec/Pim/Bundle/TransformBundle/Normalizer/Flat/ProductNormalizerSpec.php]: single quote please

File: spec/Pim/Bundle/TransformBundle/Normalizer/Flat/ProductNormalizerSpec.php [hunk 3]
```diff
@@ -72,9 +74,37 @@ function it_normalizes_product(
...
 
         $serializer->normalize($sku, 'flat', Argument::any())->willReturn(['sku' => 'sku-001']);
+        $serializer
+            ->normalize(
+                $intCodeProdValue,
+                'flat',
+                Argument::any()
+            )
+            ->willReturn([
+                12345 => "Text attribute with integer code"
```
Comment (sumbobyboys) [inline spec/Pim/Bundle/TransformBundle/Normalizer/Flat/ProductNormalizerSpec.php]: why are you using double quotes here ?

File: spec/Pim/Bundle/TransformBundle/Normalizer/Flat/ProductNormalizerSpec.php (hunk 1 above)
Comment (sumbobyboys) [inline spec/Pim/Bundle/TransformBundle/Normalizer/Flat/ProductNormalizerSpec.php]: It should return '012345' IMHO

File: src/Pim/Bundle/BaseConnectorBundle/Writer/File/CsvWriter.php [hunk 4]
```diff
@@ -215,12 +215,18 @@ protected function getAllKeys(array $items)
...
+    protected function mergeKeys(array $uniqueKeys)
     {
-        $uniqueKeys = array_fill_keys($uniqueKeys, '');
         $fullItems = [];
         foreach ($this->items as $item) {
-            $fullItems[] = array_merge($uniqueKeys, $item);
+            $filledKeys = [];
+            // We don't use array_merge here because indexes which are int val aren't cast into string and are reindexed
+            // So we loose keys for attributes which have int as code
+            foreach ($uniqueKeys as $uniqueKey) {
```
Comment (sumbobyboys) [inline src/Pim/Bundle/BaseConnectorBundle/Writer/File/CsvWriter.php]: please add behats as discussed

File: src/Pim/Bundle/BaseConnectorBundle/Writer/File/CsvWriter.php [hunk 5]
```diff
@@ -215,12 +215,18 @@ protected function getAllKeys(array $items)
...
+                $filledKeys[strval($uniqueKey)] = isset($item[$uniqueKey]) ? $item[$uniqueKey] : "";
```
Comment (sumbobyboys) [inline src/Pim/Bundle/BaseConnectorBundle/Writer/File/CsvWriter.php]: single quotes please

Comment (sumbobyboys) [inline src/Pim/Bundle/BaseConnectorBundle/Writer/File/CsvWriter.php]: simple rule: never use double quotes. Never :)

File: src/Pim/Bundle/TransformBundle/Normalizer/Flat/ProductNormalizer.php [hunk 6]
```diff
@@ -109,18 +110,23 @@ protected function normalizeProperties(ProductInterface $product)
...
+                // We don't use array_merge here because indexes which are int val aren't cast into string
+                // and are reindexed, so we loose keys for attributes which have int as code
+                foreach ($normValue as $key => $val) {
+                    $normalizedValues[strval($key)] = $val;
+                }
             }
             ksort($normalizedValues);
-            $this->results = array_merge($this->results, $normalizedValues);
+            // We don't use array_merge here because indexes which are int val aren't cast into string and are reindexed
+            foreach ($normalizedValues as $key => $normalizedValue) {
```
Comment (ClementGautier) [inline src/Pim/Bundle/TransformBundle/Normalizer/Flat/ProductNormalizer.php]: a tester mais `$this->results += $normalizedValues` devrait marcher

File: src/Pim/Bundle/BaseConnectorBundle/Writer/File/CsvWriter.php (hunk 4 above)
Comment (willy-ahva) [inline src/Pim/Bundle/BaseConnectorBundle/Writer/File/CsvWriter.php]: We have no way to check in a csv file if data are correct.

File: src/Pim/Bundle/BaseConnectorBundle/Writer/File/CsvWriter.php [hunk 7]
```diff
@@ -215,12 +215,18 @@ protected function getAllKeys(array $items)
...
     {
-        $uniqueKeys = array_fill_keys($uniqueKeys, '');
         $fullItems = [];
         foreach ($this->items as $item) {
-            $fullItems[] = array_merge($uniqueKeys, $item);
+            $filledKeys = [];
+            // We don't use array_merge here because indexes which are int val aren't cast into string and are reindexed
+            // So we loose keys for attributes which have int as code
+            foreach ($uniqueKeys as $uniqueKey) {
+                $filledKeys[strval($uniqueKey)] = isset($item[$uniqueKey]) ? $item[$uniqueKey] : '';
```
Comment (ClementGautier) [inline src/Pim/Bundle/BaseConnectorBundle/Writer/File/CsvWriter.php]: I think that `strval` here is useless.

```
php > $a = [];
php > $a['012345'] = 'a';
php > $a[12345] = 'b';
php > var_dump($a);
array(2) {
  ["012345"]=>
  string(1) "a"
  [12345]=>
  string(1) "b"
}
```

File: src/Pim/Bundle/BaseConnectorBundle/Writer/File/CsvWriter.php (hunk 4 above)
Comment (jjanvier) [inline src/Pim/Bundle/BaseConnectorBundle/Writer/File/CsvWriter.php]: possible to add behat on exports with old imports

**Pull request closed without merging**